### ISSUE TITLE
MC/Forge BucketItem can't work with modded 'waterloggable' Blocks

### ISSUE THREAD

**Issue by oOMitchOo** [MinecraftForge/MinecraftForge#6285]
**Minecraft Version:** {1.14.4}

**Forge Version:** {28.1.38}

**Steps to Reproduce:**
 1. Make your own Fluid (still, flowing, with block and bucket).
 2. Make your own Block, which should be waterloggable by that Fluid, but instead of using IWaterLoggable (which is hardcoded to work with vanilla Water) implementing IBucketPickupHandler and ILiquidContainer.
 3. Make sure ILiquidContainer.canContainFluid and ILiquidContainer.receiveFluid works with your modded Fluid.
4. Test that everything else works fine (placing the Block underwater in your Fluid)
5. Try placing your Fluid by Bucket into your waterloggable Block.
6. Nope, dis no work...

**Description of issue:**

There are two functions in the MC/Forge BucketItem, which are (partly) hardcoded to Fluids.WATER.

1. First one is in onItemRightClick, which defines a BlockPosition AFTER checking if the block at that position is ILiquidContainer AND you hold Fluids.WATER, which RESULTS in the placing position never being inside the block space, if you're holding a modded Fluid.
2. The second one is in tryPlaceContainedLiquid, which tests for ILiquidContainer AND you holding Fluids.Water BEFORE calling ILiquidContainer.receiveFluid.

So, since canContainFluid in IWaterLoggable is already hardcoded to vanilla Water and it would be great to be able to use ILiquidContainer.canContainFluid with your own Fluids,...

1. ... it would be sensible, that when the block is ILiquidContainer AND canContainFluid(the fluid you're holding), the BlockPosition would be the one of that block otherwise in front of that block.
2. ... and again, the second check here could be canContainFluid(the fluid you're holding).

**Comment by oOMitchOo** [comment]
I think @gigaherz is interested in this. Thank you!

**Comment by gigaherz** [comment]
IWaterLoggable isn't hardcoded, it has default methods which can be overriden by blocks. That said the issue still stands: BucketItem does in fact check for WATER specifically, which makes no sense.
```
BlockPos blockpos2 = blockstate.getBlock() instanceof ILiquidContainer && this.containedBlock == Fluids.WATER ? blockpos : blockpos1;
```

**Comment by gigaherz** [comment]
Fixed in a7e4cdcc (1.15.2)
